Source/Core/Core/HW/SI/SI_DeviceKeyboard.cpp

```diff
@@ -100,520 +100,115 @@ void CSIDevice_Keyboard::DoState(PointerWrap& p)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_ESC)
-  {
-    key[keys_held++] = KEY_ESC;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_INSERT)
-  {
-    key[keys_held++] = KEY_INSERT;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_DELETE)
-  {
-    key[keys_held++] = KEY_DELETE;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_TILDE)
-  {
-    key[keys_held++] = KEY_TILDE;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_BACKSPACE)
-  {
-    key[keys_held++] = KEY_BACKSPACE;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_TAB)
-  {
-    key[keys_held++] = KEY_TAB;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_CAPSLOCK)
-  {
-    key[keys_held++] = KEY_CAPSLOCK;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_LEFTSHIFT)
-  {
-    key[keys_held++] = KEY_LEFTSHIFT;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_RIGHTSHIFT)
-  {
-    key[keys_held++] = KEY_RIGHTSHIFT;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_LEFTCONTROL)
-  {
-    key[keys_held++] = KEY_LEFTCONTROL;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_RIGHTALT)
-  {
-    key[keys_held++] = KEY_RIGHTALT;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_LEFTWINDOWS)
-  {
-    key[keys_held++] = KEY_LEFTWINDOWS;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_SPACE)
-  {
-    key[keys_held++] = KEY_SPACE;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_RIGHTWINDOWS)
-  {
-    key[keys_held++] = KEY_RIGHTWINDOWS;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key4x & KEYMASK_MENU)
-  {
-    key[keys_held++] = KEY_MENU;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key5x & KEYMASK_LEFTARROW)
-  {
-    key[keys_held++] = KEY_LEFTARROW;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key5x & KEYMASK_DOWNARROW)
-  {
-    key[keys_held++] = KEY_DOWNARROW;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key5x & KEYMASK_UPARROW)
-  {
-    key[keys_held++] = KEY_UPARROW;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key5x & KEYMASK_RIGHTARROW)
-  {
-    key[keys_held++] = KEY_RIGHTARROW;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
-  if (key_status.key5x & KEYMASK_ENTER)
-  {
-    key[keys_held++] = KEY_ENTER;
-    if (keys_held >= MAX_KEYS_HELD)
-      return;
-  }
+      key[keys_held++] = keys[i];
+      if (keys_held >= MAX_KEYS_HELD)
+        return true;
```

Comment:
Wouldn't it make more sense to return false here, since this is kind of like an error?

Reply:
I used true here to represent "should we stop reading further keys", since I think it's easier to misread

`if (!thing)` compared to `if (thing)`

I can change it though if it's preferable the other way.

Reply:
ah, I guess that makes sense